src/Shape.cpp

```diff
@@ -132,44 +221,50 @@ bool Shape::contains(const Vector3& p) const {
         const Vector3 &v2 = t_points[it->i2_];
         const Vector3 &v3 = t_points[it->i3_];
 
-        Vector3 e1_U = v1 - v2;
-        Vector3 e2_U = v2 - v3;
-        Vector3 e3_U = v3 - v1;
-
-        Vector3 e1_V = v1.cross(v2);
-        Vector3 e2_V = v2.cross(v3);
-        Vector3 e3_V = v3.cross(v1);
+        geo::Line e1(v1, v2);
+        geo::Line e2(v2, v3);
+        geo::Line e3(v3, v1);
 
-        double s1 = side_operator(p_U, p_V, e1_U, e1_V);
-        double s2 = side_operator(p_U, p_V, e2_U, e2_V);
-        double s3 = side_operator(p_U, p_V, e3_U, e3_V);
+        double s1 = side_product(line, e1);
+        double s2 = side_product(line, e2);
+        double s3 = side_product(line, e3);
 
         // Determine whether v1, v2 and v3 circle line p (counter) clockwise.
-        bool clockwise = s1 < 0 && s2 < 0 && s3 < 0;
-        bool counterclockwise = s1 > 0 && s2 > 0 && s3 > 0;
+        bool clockwise = s1 >= -1e-16 && s2 >= -1e-16 && s3 >= -1e-16; // >=0
+        ushort clockwise_nz = (s1 > -1e-16) + (s2 > -1e-16) + (s3 > -1e-16); // >0
+        bool counterclockwise = s1 <= 1e-16 && s2 <= 1e-16 && s3 <= 1e-16; // <=0
+        ushort counterclockwise_nz = (s1 < 1e-16) + (s2 < 1e-16) + (s3 < 1e-16); // <0
 
-        if (clockwise || counterclockwise) { // the line passes through the triangle. now check the line segment
-            Vector3 l1_U = p_out - v1;
-            Vector3 l2_U = v1 - p;
 
-            Vector3 l1_V = p_out.cross(v1);
-            Vector3 l2_V = v1.cross(p);
+        if (!clockwise && !counterclockwise) {
+            // No intersection
+            continue;
+        }
+        else if (clockwise && counterclockwise) { // s1=0; s2=0; s3=0
+            // Coplanar
+            if (!compute_2D_intersection(Triangle(v1, v2, v3), line)) {
+                continue;
+            }
+            // Need to do something with the intersect_count
+            intersect_plane = true;
+        }
+        else if (clockwise && clockwise_nz > 0 || counterclockwise && counterclockwise_nz > 0) {
+            // Proper intersection.
+            // Now check the line segment
 
-            double s4 = side_operator(l1_U, l1_V, e2_U, e2_V);
-            double s5 = side_operator(l2_U, l2_V, e2_U, e2_V);
+            geo::Line l4(p_out, v1);
+            geo::Line l5(v1, p);
 
-            if ((s4 > 0 && s5 < 0) || (s4 < 0 && s5 > 0)) {
-                intersect_count+= counterclockwise-clockwise;
+            double s4 = side_product(l4, e2);
+            double s5 = side_product(l5, e2);
+
+            if (s4*s5>=-1e-16) {
+                intersect_count += clockwise-counterclockwise;
             }
         }
     }
 
-   if (intersect_count < 0 || intersect_count > 1) {
-        CONSOLE_BRIDGE_logError("intersect_count is %i, it should be 0 or 1! Is your shape constructed correctly?", intersect_count);
-        return false;
-    }
-
-    return intersect_count > 0;
```

Comment:
I dropped this, because when a point is on an edge of triangle and the random point outside the mesh, is on the same plane as this triangle. You can get an `abs(intersect_count)` bigger than 1.

This is the part of the PR that could be improved the most.

Reply:
How could you get a count larger than 1 though. If the line(point, outside_point) is coplanar with a triangle this does not influence the count.

Reply:
I choose to change the behaviour for this. Therefore it does indeed not happen. But when the line(point, outside_point) is coplanar, it will cross one edge twice. As the edge is part of two triangles.

But I experienced issues in an old intermediate implementation. Though I am not sure the exact behaviour anymore.